### PR TITLE
Add ObjectFile::LoadDynamicLinkingSymbolsAndUnwindRangesAsSymbols

### PR DESCRIPTION
**Pull request by dpallotti**
The choice here is to provide a single virtual method in `ObjectFile`. This is instead of providing a virtual method for loading dynamic linking symbols and a virtual method for loading unwind information ranges, and then placing the logic for merging these function at a higher level (e.g., outside of `ObjectFile`, or in a non-virtual method of `ObjectFile`). The reason is that the logic to merge the functions from these two sources is specific to the file type, as shown by the differences in implementation between `ElfFileImpl` and `CoffFileImpl`, with `CoffFileImpl::LoadDynamicLinkingSymbolsAndUnwindRangesAsSymbols` being more complicated because of missing symbol sizes from the Export Table.

Also, `CoffFileImpl::LoadSymbolsFromExportTableInternal` and `CoffFileImpl::LoadExceptionTableEntriesAsSymbolsInternal` are added to avoid reading the `RUNTIME_FUNCTION`s twice (with `GetUnwindRanges()`.

Bug: http://b/245680119

Test:
- Unit tests.
- As part of the larger symbol loading fallback prototype.

### CODE REVIEW

File: src/ObjectUtils/CoffFile.cpp [hunk 1]
```diff
@@ -58,24 +60,30 @@ class CoffFileImpl : public CoffFile {
       const llvm::object::SymbolRef& symbol_ref);
   [[nodiscard]] std::optional<SymbolInfo> CreateSymbolInfo(
       const llvm::object::SymbolRef& symbol_ref);
-  std::vector<SymbolInfo> LoadDebugSymbolsFromCoffSymbolTable(
+  [[nodiscard]] std::vector<SymbolInfo> LoadDebugSymbolsFromCoffSymbolTable(
```
Comment (florian-kuebler) [inline src/ObjectUtils/CoffFile.cpp:63]: Thanks!

File: src/ObjectUtils/CoffFile.cpp [hunk 2]
```diff
@@ -662,9 +680,71 @@ CoffFileImpl::LoadExceptionTableEntriesAsSymbols() {
...
+  ErrorMessageOr<ModuleSymbols> dynamic_linking_symbols =
+      LoadSymbolsFromExportTableInternal(unwind_ranges_or_error.value());
+  ErrorMessageOr<ModuleSymbols> unwind_ranges_as_symbols =
+      LoadExceptionTableEntriesAsSymbolsInternal(unwind_ranges_or_error.value());
+  if (!dynamic_linking_symbols.has_value() && !unwind_ranges_as_symbols.has_value()) {
+    return ErrorMessage{absl::StrFormat("%s1) %s 2) %s", kErrorMessagePrefix,
+                                        dynamic_linking_symbols.error().message(),
+                                        unwind_ranges_as_symbols.error().message())};
+  }
+  std::vector<SymbolInfo> fallback_symbols;
```
Comment (florian-kuebler) [inline src/ObjectUtils/CoffFile.cpp:709]: From the name it is a bit confusing that those are essentially the symbols for the result.

File: src/ObjectUtils/CoffFile.cpp [hunk 3]
```diff
@@ -662,9 +680,71 @@ CoffFileImpl::LoadExceptionTableEntriesAsSymbols() {
...
+      if (symbol_info.size() == 0) {
+        symbol_info.set_size(SymbolsFile::kUnknownSymbolSize);
+      }
+      fallback_symbols.emplace_back(std::move(symbol_info));
+    }
+  }
+
+  if (unwind_ranges_as_symbols.has_value()) {
+    for (SymbolInfo& symbol_info : *unwind_ranges_as_symbols.value().mutable_symbol_infos()) {
+      if (dynamic_linking_addresses.contains(symbol_info.address())) {
```
Comment (florian-kuebler) [inline src/ObjectUtils/CoffFile.cpp:728]: I might miss something, but what about the case that the dynamic linking symbol does not have a size. Can't we use the one from the  unwinding range here?

**Review by florian-kuebler — COMMENTED** [review]
Thanks for the comment. I left essentially one question here.

In general, I am not 100% sure if I agree your arguments regarding have the two implementations. In the end, the code is very very similar. Would deducing the "symbol size" hurt if we do it always?

But I am not too strong about that.

File: src/ObjectUtils/CoffFile.cpp (hunk 3 above)
Comment (dpallotti) [reply src/ObjectUtils/CoffFile.cpp:728]: Nope, we already assign the sizes from unwind info in `LoadSymbolsFromExportTableInternal`, so the exported functions that don't have a size yet are exactly for which we don't have unwind info. This is basically explained in the comment above. It's the leaf functions in the Microsoft definition.

**Review by dpallotti — COMMENTED** [review]
> In general, I am not 100% sure if I agree your arguments regarding have the two implementations. In the end, the code is very very similar. Would deducing the "symbol size" hurt if we do it always?

Originally I had a single implementation as the alternative described in the commit message:
- While the deductions of sizes might not hurt for ELF files, that's not what we want to do and it feels wrong to do it no matter what. So I had `if`s checking for `IsCoff()`. The code was not super nice but perhaps acceptable.
- Then I wanted to avoid calling `GetUnwindRanges` twice, once in `LoadSymbolsFromExportTable` and once in `LoadExceptionTableEntriesAsSymbols`. Hence the addition and usage of the `Internal` versions. That simply requires a separate function.

So the logic is very similar, but the code not so much. Considering the common code is fairly short and straightforward (the two for loops, so like ten lines of code), I find that trying to reuse it is not worth it.

File: src/ObjectUtils/CoffFile.cpp (hunk 3 above)
Comment (florian-kuebler) [reply src/ObjectUtils/CoffFile.cpp:728]: Okay. Thanks for the refresher!

**Review by florian-kuebler — APPROVED** [review]
LGTM. Thanks for the additional input!

**Comment by dpallotti** [comment]
> Thanks for the additional input!

Thank you for the review!

**Pull request merged**